pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -0,0 +1,346 @@
+		},
+	}
+
+	cmd.Flags().Bool("all", false, "Build dependency trees for all image repositories")
+	cmd.Flags().StringP("output", "o", "json", "Output format of dependency tree(s)")
+	return cmd
+}
+
+// getRepos iterates over a given set of build configurations
+// and extracts all the image repositories on which other
+// image repositories are based
+func getRepos(configs []buildapi.BuildConfig, defaultNamespace string) map[string]string {
+	avoidDuplicates := make(map[string]string)
+	for _, cfg := range configs {
+		for _, tr := range cfg.Triggers {
+			if tr.ImageChange != nil {
+				switch tr.ImageChange.From.Namespace {
+				case "":
+					avoidDuplicates[tr.ImageChange.From.Name] = defaultNamespace
+				default:
+					avoidDuplicates[tr.ImageChange.From.Name] = tr.ImageChange.From.Namespace
+				}
+			}
+		}
+	}
+
+	return avoidDuplicates
+}
+
+// breakDown accepts the input passed by the user and
+// breaks it down to the desired values if possible
+func breakDown(input string) (string, string, error) {
+	args := strings.Split(input, ":")
+	switch len(args) {
+	case 1:
+		return args[0], "latest", nil
+	case 2:
+		if strings.TrimSpace(args[1]) == "" {
+			return args[0], "latest", nil
+		}
+		return args[0], args[1], nil
+	default:
+		return "", "", invalidRepoTagErr
+	}
+}
+
+// flushMap deletes all they keys from a map
+func flushMap(repoList map[string]string) {
+	for key := range repoList {
+		delete(repoList, key)
+	}
+}
+
+// findRepoDeps accepts an image repository name and a list of build
+// configurations and returns the dependency tree of the specified
+// image repository
+func findRepoDeps(name, namespace, tag string, all bool, buildConfigList []buildapi.BuildConfig) *ImageRepo {
+	root := &ImageRepo{
+		Name:      name,
+		Namespace: namespace,
+		Tags:      []string{tag},
+	}
+
+	// Search all build configurations in order to find the image
+	// repositories depending on the specified image repository
+	var outputName, outputTag string
+	for _, cfg := range buildConfigList {
+		for _, tr := range cfg.Triggers {
+			// If an ImageChangeTrigger has been specified (ie. it's not nil), use its From
+			// and Tag fields to compare against the specified image repository
+			if tr.ImageChange != nil && name == tr.ImageChange.From.Name && (all || tag == tr.ImageChange.Tag) {
+				// Either To & Tag or DockerImageReference will be used as output
+				if cfg.Parameters.Output.To.Name != "" {
+					outputName = cfg.Parameters.Output.To.Name
+					outputTag = cfg.Parameters.Output.Tag
+					if cfg.Parameters.Output.To.Namespace != "" {
+						namespace = cfg.Parameters.Output.To.Namespace
+					}
+				} else {
+					ref, err := imageapi.ParseDockerImageReference(cfg.Parameters.Output.DockerImageReference)
+					checkErr(err)
+					outputName = ref.Name
+					outputTag = ref.Tag
+				}
+
+				// Build all children and their dependency trees recursively
+				child := findRepoDeps(outputName, namespace, outputTag, all, buildConfigList)
```

Comment:
It'd be good to keep a list of parents that you pass to findRepoDeps to avoid a cycle. Only add children that are not in the list of parents.